src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUseageDescriptorFactory.cs

```diff
@@ -0,0 +1,136 @@
+        private static readonly char[] InvalidPathCharacters = Path.GetInvalidPathChars();
+
+        public static TagHelperUseageDescriptor CreateDescriptor([NotNull] TypeInfo typeInfo)
+        {
+            var id = XmlDocumentationProvider.GetId(typeInfo);
+            return CreateDescriptorCore(typeInfo.Assembly, id);
+        }
+
+        public static TagHelperUseageDescriptor CreateDescriptor([NotNull] PropertyInfo propertyInfo)
+        {
+            var id = XmlDocumentationProvider.GetId(propertyInfo);
+            var declaringTypeAssembly = propertyInfo.DeclaringType.GetTypeInfo().Assembly;
+            return CreateDescriptorCore(declaringTypeAssembly, id);
+        }
+
+        private static TagHelperUseageDescriptor CreateDescriptorCore(Assembly typeAssembly, string id)
+        {
+            var typeAssemblyLocation = typeAssembly.Location;
+
+            if (string.IsNullOrWhiteSpace(typeAssemblyLocation) && !string.IsNullOrWhiteSpace(typeAssembly.CodeBase))
+            {
+                var uri = new UriBuilder(typeAssembly.CodeBase);
+
+                // Normalize the path to a UNC path. This will remove things like file:// from start of the uri.Path.
+                typeAssemblyLocation = Uri.UnescapeDataString(uri.Path);
+
+                // Still couldn't resolve a valid typeAssemblyLocation.
+                if (string.IsNullOrWhiteSpace(typeAssemblyLocation))
+                {
+                    return null;
+                }
+            }
+
+            XmlDocumentationProvider documentationProvider = null;
+            if (!XmlDocumentationProviderCache.TryGetValue(typeAssemblyLocation, out documentationProvider))
+            {
+                var xmlDocumentationFile = GetXmlDocumentationFile(typeAssemblyLocation);
+
+                // We only want to process the file if it exists. In the case it doesn't, a null value will be added
+                // to the cache to not constantly look for new XML files.
+                if (xmlDocumentationFile != null)
+                {
+                    documentationProvider = new XmlDocumentationProvider(xmlDocumentationFile.FullName);
+                }
+
+                XmlDocumentationProviderCache.Add(typeAssemblyLocation, documentationProvider);
+            }
+
+            // Members will be null if there is no associated XML file for the provided typeAssembly.
+            if (documentationProvider != null)
+            {
+                if (documentationProvider.HasDocumentation(id))
+                {
+                    var summary = documentationProvider.GetSummary(id);
+                    var remarks = documentationProvider.GetRemarks(id);
+
+                    return new TagHelperUseageDescriptor(summary, remarks);
+                }
+            }
+
+            return null;
+        }
+
+        private static FileInfo GetXmlDocumentationFile(string typeAssemblyLocation)
```

Comment:
Debated moving the functionality of this method into another type such as `XmlDocumentationLocator` but thought it might be overkill for now.